core/domain/topic_services.py

```diff
@@ -1814,32 +1833,38 @@ def get_all_published_story_exploration_ids(
         [topic_fetchers.get_topic_summary_by_id(topic_id)] if topic_id
         else topic_fetchers.get_all_topic_summaries())
 
-    # Keep all topic summaries with a mapping. For those without a mapping,
+    # Keep each summary's mapping. For those without a mapping,
     # record their ids, fetch their corresponding topics with them, and then
-    # use the topic to compute their summary. Add each new topic summary to
-    # the list of summaries that initially had a mapping.
-    topic_summaries_with_persisted_mapping = []
+    # use the topic to compute the mapping. Add each computed mapping to
+    # the list of persisted mappings.
+    mappings = []
     ids_of_summaries_without_persisted_mapping: List[str] = []
     for summary in fetched_topic_summaries:
         if summary.published_story_exploration_mapping is None:
             ids_of_summaries_without_persisted_mapping.append(summary.id)
         else:
-            topic_summaries_with_persisted_mapping.append(summary)
-    topic_summaries_with_recently_computed_mapping = []
+            mappings.append(
+                summary.published_story_exploration_mapping)
     if len(ids_of_summaries_without_persisted_mapping) > 0:
         topics = topic_fetchers.get_topics_by_ids(
             ids_of_summaries_without_persisted_mapping)
         for topic in topics:
             assert topic is not None
-            topic_summaries_with_recently_computed_mapping.append(
-                compute_summary_of_topic(topic))
-    topic_summaries = (
-        topic_summaries_with_persisted_mapping +
-        topic_summaries_with_recently_computed_mapping)
+            published_story_ids = [
+                story_ref.story_id for story_ref
+                in topic.canonical_story_references +
+                    topic.additional_story_references
+                if story_ref.story_is_published
+            ]
+            published_stories: List[story_domain.Story] = [
+                story for story in story_fetchers.get_stories_by_ids(
```

Comment:
Can you do this in a way that fetches all stories in a single database call, rather than N calls? Maybe gather all the story IDs first and convert them all to stories later?

If it makes things complicated, then an alternative (since we don't have that many stories) might be to just do a get_all() to get every single story at the outset.

Reply:
Done. I did the call first and converted them later.